src/main/java/algorithms/curated170/medium/TreeDiameter.java

```diff
@@ -0,0 +1,56 @@
+package algorithms.curated170.medium;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class TreeDiameter {
+      private int diameter = 0;
+      private Map<Integer, List<Integer>> tree = new HashMap<>();
+
+      public int treeDiameter(int[][] input) {
+            int ret = 0;
+            for (int[] pair : input) {
+                  int a = pair[0];
+                  int b = pair[1];
+                  tree.putIfAbsent(a, new ArrayList<>());
+                  tree.putIfAbsent(b, new ArrayList<>());
+                  tree.get(a).add(b);
+                  tree.get(b).add(a);
+            }
+            searchDiameter(tree.keySet().iterator().next(), Integer.MIN_VALUE);
+            return diameter;
+      }
+
+      public int searchDiameter(int root, int parent) {
+            int maxArm1 = 0;
+            int maxArm2 = 0;
+            for (int k : tree.get(root)) {
+                  int depth = 0;
+                  if (k != parent) {
```

Comment:
**k!=parent**:
A nice alternative to using **boolean visited[]** to prevent circle. As there is no duplicate value; we don't need an extra array.